godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,638 @@
+        // deprecated by Godot
+        notification_func: None,
+        to_string_func: Some(script_instance_info::to_string_func::<T>),
+        refcount_decremented_func: Some(script_instance_info::refcount_decremented_func::<T>),
+        refcount_incremented_func: Some(script_instance_info::refcount_incremented_func::<T>),
+        is_placeholder_func: Some(script_instance_info::is_placeholder_func::<T>),
+        // usage is unclear
+        set_fallback_func: None,
+    };
+
+    let instance_ptr = Box::into_raw(Box::new(gd_instance));
+
+    let data = ScriptInstanceData {
+        inner: RefCell::new(rs_instance),
+        gd_instance_ptr: instance_ptr,
+    };
+
+    let data_ptr = Box::into_raw(Box::new(data));
+
+    // SAFETY: `script_instance_create` expects a `GDExtensionScriptInstanceInfoPtr` and a
+    // generic `GDExtensionScriptInstanceDataPtr` of our choice. The validity of the
+    // instance info struct is ensured by code generation.
+    unsafe {
+        get_interface().script_instance_create.unwrap()(
+            instance_ptr,
+            data_ptr as GDExtensionScriptInstanceDataPtr,
+        )
+    }
+}
+
+mod script_instance_info {
+    use std::cell::{BorrowError, Ref, RefMut};
+    use std::ffi::c_void;
+    use std::mem::{self, ManuallyDrop};
+    use std::ptr::drop_in_place;
+
+    use crate::builtin::meta::GodotType;
+    use crate::builtin::{GString, StringName, Variant};
+    use crate::engine::{Object, ScriptLanguage};
+    use crate::obj::Gd;
+    use crate::sys::{
+        GDExtensionBool, GDExtensionCallError, GDExtensionConstStringNamePtr,
+        GDExtensionConstVariantPtr, GDExtensionInt, GDExtensionMethodInfo, GDExtensionObjectPtr,
+        GDExtensionPropertyInfo, GDExtensionScriptInstanceDataPtr, GDExtensionScriptLanguagePtr,
+        GDExtensionStringPtr, GDExtensionTypePtr, GDExtensionVariantPtr, GDExtensionVariantType,
+        GodotFfi, PtrcallType, GDEXTENSION_CALL_OK,
+    };
+
+    use super::{ScriptInstance, ScriptInstanceData};
+
+    fn borrow_instance_mut<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> RefMut<'a, T> {
+        unsafe { (*instance).inner.borrow_mut() }
+    }
+
+    fn borrow_instance<'a, T: ScriptInstance>(instance: *mut ScriptInstanceData<T>) -> Ref<'a, T> {
+        unsafe { (*instance).inner.borrow() }
+    }
+
+    fn try_borrow_instance<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> Result<Ref<'a, T>, BorrowError> {
+        unsafe { (*instance).inner.try_borrow() }
+    }
+
+    /// # Safety
+    /// - we expect the engine to provide a vallid instance pointer
+    fn instance_data_as_script_instance<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+    ) -> *mut ScriptInstanceData<T> {
+        p_instance as *mut ScriptInstanceData<T>
+    }
+
+    /// # Safety
+    /// - we expect the engine to provide a valid const string name pointer
```

Comment:
```suggestion
    /// # Safety
    /// We expect the engine to provide a valid const string name pointer.
```

Reply:
I would rephrase this and any other similar functions as `p_name must be a valid const string name pointer.`. And then when we hand off these callbacks to the engine we later state that the engine is expected to uphold the safety invariants of the functions.